olive/passes/onnx/gptq.py

```diff
@@ -0,0 +1,200 @@
+
+    @torch.no_grad()
+    def _run_for_config(
+        self, model: PyTorchModelHandler, data_root: str, config: Dict[str, Any], output_model_path: str
+    ) -> ONNXModelHandler:
+        from onnxruntime import __version__ as ort_version
+        from optimum.exporters.onnx import onnx_export_from_model
+        from optimum.gptq import GPTQQuantizer
+        from optimum.version import __version__ as optimum_version
+        from transformers import AutoTokenizer
+
+        from olive.passes.onnx.gptq_utils import QuantLinearORT
+
+        if version.parse(optimum_version) < version.parse("1.17.0"):
+            raise ValueError("Please use optimum>=1.17.0 for gptq quantization")
+        if version.parse(ort_version) < version.parse("1.17.0"):
+            raise ValueError("Please use onnxruntime-gpu>=1.17.0 for gptq quantization")
+        if (
+            self.accelerator_spec.accelerator_type != Device.GPU
+            or self.accelerator_spec.execution_provider != "CUDAExecutionProvider"
+        ):
+            raise ValueError("Please use GPU and CUDAExecutionProvider to run gptq quantization.")
+
+        tokenizer = AutoTokenizer.from_pretrained(model.hf_config.model_name, use_fast=False)
+
+        dataset = None
+        if config["dataloader_func"]:
+            dataset = self._user_module_loader.call_object(
+                config["dataloader_func"],
+                **(config["dataloader_func_kwargs"] or {}),
+            )
+        elif config["data_config"]:
+            data_config = validate_config(config["data_config"], DataConfig)
+            dataloader = data_config.to_data_container().create_dataloader(data_root)
+            dataset = [data[0] for data in dataloader]
+        else:
+            dataset = config["dataset"]
+            if not dataset:
+                raise ValueError("Please provide dataloader_func, data_config or dataset for gptq quantization.")
+
+        pytorch_model = model.load_model()
+        quantizer = GPTQQuantizer(
+            bits=config["bits"],
+            dataset=dataset,
+            block_name_to_quantize=config["block_name_to_quantize"],
+            group_size=config["group_size"],
+            damp_percent=config["damp_percent"],
+            static_groups=config["static_groups"],
+            true_sequential=config["true_sequential"],
+            desc_act=config["desc_act"],
+            sym=config["sym"],
+        )
+
+        def get_onnx_quant_linear(*args, **kwargs):
+            return QuantLinearORT
+
+        import optimum
+
+        original = optimum.gptq.quantizer.dynamically_import_QuantLinear
+        try:
+            # GPTQ Quantization in transformers use optimum under the hood, so we
+            # replace QuantLinear in optimum with QuantLinearORT for quant linear layer packing
+            optimum.gptq.quantizer.dynamically_import_QuantLinear = get_onnx_quant_linear
+            quantized_model = quantizer.quantize_model(pytorch_model, tokenizer)
```

Comment:
>quantized_model = quantizer.quantize_model(pytorch_model, tokenizer) [](http://example.com/codeflow?start=12&length=68)

add comments about how the underlying device is used.
For e.g. force move the layer to cuda and when quantization finish, move it back to cpu by default.